### PR TITLE
WIP: Renamer system rewrite

### PR DESCRIPTION
**Pull request by revam**

### CODE REVIEW

File: Shoko.Server/Renamer/RenameFileHelper.cs [hunk 1]
```diff
@@ -172,11 +180,17 @@ internal static void FindRenamers(IList<Assembly> assemblies)
...
-                            $"[RENAMER] Warning Duplicate renamer key \"{key}\" of types {implementation}@{implementation.Assembly.Location} and {Renamers[key]}@{Renamers[key].type.Assembly.Location}");
+                            $"[RENAMER] Warning Duplicate renamer key \"{key}\" of types {implementation}@{implementation.Assembly.Location} and {Renamers[key]}@{Renamers[key].Type.Assembly.Location}");
                         continue;
                     }
 
-                    Renamers.Add(key, (implementation, desc));
+                    Renamers.Add(key, new RenamerInfo
+                    {
+                        ID = Renamers.Count + 1,
+                        Name = key,
```
Comment (Cazzar) [inline Shoko.Server/Renamer/RenameFileHelper.cs:190]: You're double storing key here, which is pointless.

Renamer ID should never be dynamically generated as it will be confusing to runtime, what is even the reason, you mentioned in discord "Comprimises" but I can't think of anything at this point in time.

File: Shoko.Server/Renamer/RenameFileHelper.cs [hunk 2]
```diff
@@ -20,8 +20,16 @@ namespace Shoko.Server
 {
     public class RenameFileHelper
     {
+        public class RenamerInfo
+        {
+            public int ID { get; set; }
+            public string Name { get; set; }
+            public string Description { get; set; }
+            public Type Type { get; set; }
+        }
```
Comment (Cazzar) [inline Shoko.Server/Renamer/RenameFileHelper.cs:29]: What is even the reason for changing how the renamer info is stored- to get extra data? But the only thing there is added is a false "ID" which is already defined by the IDictonary key.

File: Shoko.Server/Models/SVR_VideoLocal_Place.cs [hunk 3]
```diff
@@ -398,15 +398,15 @@ public bool RefreshMediaInfo()
             }
         }
         
-        public void RemoveRecordAndDeletePhysicalFile(bool deleteFolder = true)
+        public bool RemoveRecordAndDeletePhysicalFile(bool deleteFolder = true)
```
Comment (Cazzar) [inline Shoko.Server/Models/SVR_VideoLocal_Place.cs:401]: What are all these changes for, if you are wanting to add renamer visibility, this seems to be a lot of irrelevant changes.

File: Shoko.Server/Models/SVR_VideoLocal_Place.cs [hunk 4]
```diff
@@ -469,6 +470,65 @@ public void RemoveAndDeleteFileWithOpenTransaction(ISession session, HashSet<SVR
...
         }
 
+        public class RelocateResult
+        {
+            public bool Success { get; set; }
+
+            public string RelativePath { get; set; }
+
+            public SVR_ImportFolder ImportFolder { get; set; }
+        }
```
Comment (Cazzar) [inline Shoko.Server/Models/SVR_VideoLocal_Place.cs:480]: This entire implementation could be a tuple return - but again this is seeming to be a lot of unrelated changes.

File: Shoko.Server/API/v3/Controllers/FileController.cs [hunk 5]
```diff
@@ -330,6 +330,7 @@ private ActionResult ScrobbleStatusOnFile(SVR_VideoLocal file, bool? watched, lo
         /// <param name="fileID">VideoLocal ID</param>
         /// <param name="value">Thew new ignore value.</param>
         /// <returns></returns>
+        [Authorize("admin")]
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/FileController.cs:333]: Unrelated change?

File: Shoko.Server/API/v3/Controllers/FileController.cs [hunk 6]
```diff
@@ -348,6 +349,7 @@ public ActionResult IgnoreFile([FromRoute] int fileID, [FromQuery] bool value =
         /// </summary>
         /// <param name="fileID">VideoLocal ID</param>
         /// <returns></returns>
+        [Authorize("admin")]
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/FileController.cs:352]: Unrelated change?

File: Shoko.Server/API/v3/Controllers/FileController.cs [hunk 7]
```diff
@@ -399,6 +402,7 @@ public ActionResult RescanFile([FromRoute] int fileID, [FromQuery] bool priority
...
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/FileController.cs:405]: Unrelated change?

File: Shoko.Server/API/v3/Controllers/FileController.cs [hunk 8]
```diff
@@ -422,6 +426,7 @@ public ActionResult RehashFile([FromRoute] int fileID)
         /// <param name="fileID">The file id.</param>
         /// <param name="body">The body.</param>
         /// <returns></returns>
+        [Authorize("admin")]
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/FileController.cs:429]: Unrelated change?

File: Shoko.Server/API/v3/Controllers/FileController.cs [hunk 9]
```diff
@@ -638,6 +748,7 @@ public ActionResult LinkMultipleFiles([FromBody] File.Input.LinkSeriesMultipleBo
         /// </summary>
         /// <param name="body">The body.</param>
         /// <returns></returns>
+        [Authorize("admin")]
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/FileController.cs:751]: Unrelated change?

File: Shoko.Server/API/v3/Controllers/RenamerController.cs [hunk 10]
```diff
@@ -13,39 +14,204 @@ namespace Shoko.Server.API.v3.Controllers
...
-        [HttpDelete("{renamerID}")]
-        public ActionResult Disable(string renamerID)
+        /// <summary>
+        /// Enable the <see cref="Models.Shoko.Renamer"/> by the given <paramref name="renamerID"/>.
+        /// </summary>
+        /// <param name="renamerID">Renamer ID</param>
+        /// <returns></returns>
+        [Authorize("admin")]
+        [HttpPost("{renamerID}/Enable")]
+        public ActionResult EnableRenamerByRenamerID(int renamerID)
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/RenamerController.cs:70]: What does the int ID relate to?
Renamer types are stored by a string key, which is how they are defined in software for consistency reasons.

File: Shoko.Server/API/v3/Controllers/RenamerController.cs [hunk 11]
```diff
@@ -13,39 +14,204 @@ namespace Shoko.Server.API.v3.Controllers
...
         {
-            ServerSettings.Instance.Plugins.EnabledRenamers[renamerID] = false;
+            var renamer = RenameFileHelper.Renamers.Values.FirstOrDefault(r => r.ID == renamerID);
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/RenamerController.cs:72]: String key lookup

File: Shoko.Server/API/v3/Controllers/RenamerController.cs [hunk 12]
```diff
@@ -13,39 +14,204 @@ namespace Shoko.Server.API.v3.Controllers
...
+            if (renamer == null)
+                return NotFound("Renamer not found.");
+
+            ServerSettings.Instance.Plugins.EnabledRenamers[renamer.Name] = true;
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/RenamerController.cs:76]: IOOB issue

File: Shoko.Server/API/v3/Controllers/RenamerController.cs [hunk 13]
```diff
@@ -13,39 +14,204 @@ namespace Shoko.Server.API.v3.Controllers
...
-        public ActionResult SetPriority(string renamerID, [FromBody] int priority)
+        /// <summary>
+        /// Set the priority of the <see cref="Models.Shoko.Renamer"/> by the given <paramref name="renamerID"/>, optionally enabling it if it's disabled.
+        /// </summary>
+        /// <param name="renamerID">Renamer ID</param>
+        /// <param name="priority">Renamer priority. The lower number the higher the priority.</param>
+        /// <returns></returns>
+        [Authorize("admin")]
+        [HttpPost("{renamerID}/SetPriority")]
+        public ActionResult EditRenamerByRenamerID(int renamerID, [FromQuery] [Range(0, int.MaxValue)] int priority)
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/RenamerController.cs:90]: I am looking soon to change fundamentally how we pipeline renamers/scripts, as mentioned in discord

File: Shoko.Server/API/v3/Controllers/RenamerController.cs [hunk 14]
```diff
@@ -13,39 +14,204 @@ namespace Shoko.Server.API.v3.Controllers
...
+        /// Add a new script.
+        /// </summary>
+        /// <param name="body">The script to add.</param>
+        /// <returns></returns>
+        [Authorize("admin")]
+        [HttpPost("Script")]
+        public ActionResult AddRenamerScript([FromBody] Models.Shoko.Renamer.Script body)
+        {
+            if (!body.IDs.Renamer.HasValue)
+                return BadRequest("The renamer id is missing! Set Renamer.Script.IDs.Renamer to a valid renamer id!");
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/RenamerController.cs:141]: Reamer ID should always be a string, integer IDs here do not make sense.

File: Shoko.Server/API/v3/Controllers/RenamerController.cs [hunk 15]
```diff
@@ -13,39 +14,204 @@ namespace Shoko.Server.API.v3.Controllers
...
+        /// Get a <see cref="Models.Shoko.Renamer.Script"/> by the given <paramref name="scriptID"/>.
+        /// </summary>
+        /// <param name="scriptID">Script ID</param>
+        /// <returns>The script</returns>
+        [HttpGet("Script/{scriptID}")]
+        public ActionResult<Models.Shoko.Renamer.Script> GetRenamerScriptByScriptID([FromRoute] int scriptID)
+        {
+            var script = RepoFactory.RenameScript.GetByID(scriptID);
+            if (script == null)
+                return NotFound("Renamer.Script not found.");
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/RenamerController.cs:174]:
```suggestion
                return NotFound("Renamer Script not found.");
```

File: Shoko.Server/API/v3/Controllers/RenamerController.cs [hunk 16]
```diff
@@ -13,39 +14,204 @@ namespace Shoko.Server.API.v3.Controllers
...
+        /// <param name="scriptID">Script ID</param>
+        /// <param name="body">The new script to replace the entry with.</param>
+        /// <returns></returns>
+        [Authorize("admin")]
+        [HttpPut("Script/{scriptID}")]
+        public ActionResult RutRenamerScriptByScriptID([FromRoute] int scriptID, [FromBody] Models.Shoko.Renamer.Script body)
+        {
+            var script = RepoFactory.RenameScript.GetByID(scriptID);
+            if (script == null)
+                return NotFound("Renamer.Script not found.");
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/RenamerController.cs:191]:
```suggestion
                return NotFound("Renamer Script not found.");
```

File: Shoko.Server/API/v3/Controllers/RenamerController.cs [hunk 17]
```diff
@@ -13,39 +14,204 @@ namespace Shoko.Server.API.v3.Controllers
...
+
+            body.MergeWithExisting(script);
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/RenamerController.cs:193]: Docblock/method indicates a replace, but this is saying merge? Possible bug?

File: Shoko.Server/API/v3/Controllers/RenamerController.cs [hunk 18]
```diff
@@ -13,39 +14,204 @@ namespace Shoko.Server.API.v3.Controllers
...
+        /// </summary>
+        /// <param name="scriptID">Script ID</param>
+        /// <returns></returns>
+        [Authorize("admin")]
+        [HttpDelete("Script/{scriptID}")]
+        public ActionResult DeleteRenamerScriptByScriptID([FromRoute] int scriptID)
+        {
+            var script = RepoFactory.RenameScript.GetByID(scriptID);
+            if (script == null)
+                return NotFound("Renamer.Script not found.");
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Controllers/RenamerController.cs:210]:
```suggestion
                return NotFound("Renamer Script not found.");
```

File: Shoko.Server/API/v3/Models/Shoko/File.cs [hunk 19]
```diff
@@ -105,17 +105,18 @@ public class Location
             /// <summary>
             /// The Import Folder that this file resides in 
             /// </summary>
+            [Required]
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Models/Shoko/File.cs:108]: Unrelated change?

File: Shoko.Server/API/v3/Models/Shoko/File.cs [hunk 20]
```diff
@@ -105,17 +105,18 @@ public class Location
...
             public int ImportFolderID { get; set; }
             
             /// <summary>
             /// The relative path from the import folder's path on the server. The Filename can be easily extracted from this. Using the ImportFolder, you can get the full server path of the file or map it if the client has remote access to the filesystem. 
             /// </summary>
+            [Required]
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Models/Shoko/File.cs:114]: Unrelated change?

File: Shoko.Server/API/v3/Models/Shoko/File.cs [hunk 21]
```diff
@@ -105,17 +105,18 @@ public class Location
...
             public string RelativePath { get; set; }
             
             /// <summary>
             /// Can the server access the file right now
             /// </summary>
-            [JsonRequired]
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Models/Shoko/File.cs:118]: Unrelated change?

File: Shoko.Server/API/v3/Models/Shoko/Renamer.cs [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+using System.Linq;
+
+namespace Shoko.Server.API.v3.Models.Shoko
+{
+    public class Renamer : BaseModel
+    {
+        public Renamer(RenameFileHelper.RenamerInfo renamerInfo)
+        {
+            var scripts = RepoFactory.RenameScript.GetRenamerType(renamerInfo.Name);
+            ID = renamerInfo.ID;
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Models/Shoko/Renamer.cs:20]: Int IDs for renamers faked in make no sense, and will cause possible client side cache bugs

File: Shoko.Server/API/v3/Models/Shoko/Renamer.cs [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+                Body = script.Script;
+            }
+
+            public void MergeWithExisting(RenameScript existing)
+            {
+                if (IDs.Renamer.HasValue)
+                {
+                    var renamerInfo = RenameFileHelper.Renamers.Values.FirstOrDefault(r => r.ID == IDs.Renamer.Value);
+                    if (renamerInfo != null)
+                        existing.RenamerType = renamerInfo.Name;
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Models/Shoko/Renamer.cs:73]: Why pull it from an arbitrary name when the string type is what was already used?

File: Shoko.Server/API/v3/Models/Shoko/Renamer.cs [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+                existing.Script = Body;
+                existing.ScriptName = Name;
+                existing.IsEnabledOnImport = EnabledOnImport ? 1 : 0;
+            }
+
+            /// <summary>
+            /// Script IDs.
+            /// </summary>
+            [Required]
+            public RenamerScriptIDs IDs { get; set; }
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Models/Shoko/Renamer.cs:84]: I understand this is a thing in APIv3 as a whole, but the entire "IDs" concept seems to be over-complicated, or even over used.

File: Shoko.Server/API/v3/Models/Shoko/Renamer.cs [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+            /// Determines if the script should ran automatically on import if the renamer is enabled.
+            /// </summary>
+            [Required]
+            public bool EnabledOnImport { get; set; }
+
+            /// <summary>
+            /// The script body.
+            /// </summary>
+            [Required(AllowEmptyStrings = true)]
+            public string Body { get; set; }
```
Comment (Cazzar) [inline Shoko.Server/API/v3/Models/Shoko/Renamer.cs:107]: This should be nullable, and not required.

File: Shoko.Server/Repositories/Direct/RenameScriptRepository.cs [hunk 26]
```diff
@@ -40,5 +40,17 @@ public RenameScript GetByName(string scriptName)
                 .List<RenameScript>().FirstOrDefault();
             return cr;
         }
+
+        public List<RenameScript> GetRenamerType(string renamerType)
+        {
+            if (string.IsNullOrEmpty(renamerType)) return new();
```
Comment (Cazzar) [inline Shoko.Server/Repositories/Direct/RenameScriptRepository.cs:46]: Why are we returning a new list if there is no type?

**Review by Cazzar — CHANGES_REQUESTED** [review]

File: Shoko.Server/API/v3/Controllers/FileController.cs (hunk 5 above)
Comment (revam) [reply Shoko.Server/API/v3/Controllers/FileController.cs:333]: Yes.

File: Shoko.Server/API/v3/Controllers/FileController.cs (hunk 6 above)
Comment (revam) [reply Shoko.Server/API/v3/Controllers/FileController.cs:352]: Yes.

File: Shoko.Server/API/v3/Controllers/FileController.cs [hunk 27]
```diff
@@ -377,6 +379,7 @@ public ActionResult<AVDumpResult> AvDumpFile([FromRoute] int fileID)
         /// <param name="fileID">VideoLocal ID</param>
         /// <param name="priority">Increase the priority to the max for the queued command.</param>
         /// <returns></returns>
+        [Authorize("admin")]
```
[the remark this answers is not in the document]
Comment (revam) [reply Shoko.Server/API/v3/Controllers/FileController.cs:382]: Yes.

File: Shoko.Server/API/v3/Controllers/FileController.cs (hunk 7 above)
Comment (revam) [reply Shoko.Server/API/v3/Controllers/FileController.cs:405]: Yes.

File: Shoko.Server/API/v3/Controllers/FileController.cs (hunk 8 above)
Comment (revam) [reply Shoko.Server/API/v3/Controllers/FileController.cs:429]: Yes.

File: Shoko.Server/API/v3/Controllers/FileController.cs [hunk 28]
```diff
@@ -445,11 +450,114 @@ public ActionResult LinkSingleEpisodeToFile([FromRoute] int fileID, [FromBody] F
...
+                RelativePath = place.FilePath,
+            };
+        }
+
         /// <summary>
         /// Link one or more episodes from a series to the same file.
         /// </summary>
         /// <param name="fileID">The file id.</param>
         /// <param name="body">The body.</param>
+        [Authorize("admin")]
```
[the remark this answers is not in the document]
Comment (revam) [reply Shoko.Server/API/v3/Controllers/FileController.cs:560]: Yes.

File: Shoko.Server/API/v3/Controllers/FileController.cs [hunk 29]
```diff
@@ -514,6 +622,7 @@ public ActionResult LinkMultipleEpisodesToFile([FromRoute] int fileID, [FromBody
         /// <param name="fileID">The file id.</param>
         /// <param name="body">Optional. The body.</param>
         /// <returns></returns>
+        [Authorize("admin")]
```
[the remark this answers is not in the document]
Comment (revam) [reply Shoko.Server/API/v3/Controllers/FileController.cs:625]: Yes.

File: Shoko.Server/API/v3/Controllers/FileController.cs [hunk 30]
```diff
@@ -554,6 +663,7 @@ public ActionResult UnlinkMultipleEpisodesFromFile([FromRoute] int fileID, [From
         /// </summary>
         /// <param name="body">The body.</param>
         /// <returns></returns>
+        [Authorize("admin")]
```
[the remark this answers is not in the document]
Comment (revam) [reply Shoko.Server/API/v3/Controllers/FileController.cs:666]: Yes.

File: Shoko.Server/API/v3/Controllers/FileController.cs (hunk 9 above)
Comment (revam) [reply Shoko.Server/API/v3/Controllers/FileController.cs:751]: Yes.

File: Shoko.Server/API/v3/Controllers/RenamerController.cs (hunk 16 above)
Comment (revam) [reply Shoko.Server/API/v3/Controllers/RenamerController.cs:191]: It's intentional. The model name is "Renamer.Script" in the api.

File: Shoko.Server/API/v3/Controllers/RenamerController.cs (hunk 18 above)
Comment (revam) [reply Shoko.Server/API/v3/Controllers/RenamerController.cs:210]: It's intentional. The model name is "Renamer.Script" in the api.

File: Shoko.Server/API/v3/Models/Shoko/File.cs (hunk 20 above)
Comment (revam) [reply Shoko.Server/API/v3/Models/Shoko/File.cs:114]: No. It's for submitting file locations when relocating files.

File: Shoko.Server/API/v3/Models/Shoko/File.cs (hunk 21 above)
Comment (revam) [reply Shoko.Server/API/v3/Models/Shoko/File.cs:118]: No. It's for submitting file locations when relocating files. Also with the new default of including every field value by default then I don't see why this attribute is still required as it's always sent from the server-side and can be omitted when used in a body sent from the client-side (since it have no effect on the server).

File: Shoko.Server/API/v3/Models/Shoko/Renamer.cs (hunk 25 above)
Comment (revam) [reply Shoko.Server/API/v3/Models/Shoko/Renamer.cs:107]: Fixed.

File: Shoko.Server/Models/SVR_VideoLocal_Place.cs (hunk 3 above)
Comment (revam) [reply Shoko.Server/Models/SVR_VideoLocal_Place.cs:401]: The PR is for adding everything related to the renamers — including file relocation — and not just renamer viability, hench why these changes and also the file endpoints are in the PR.

File: Shoko.Server/Models/SVR_VideoLocal_Place.cs (hunk 4 above)
Comment (revam) [reply Shoko.Server/Models/SVR_VideoLocal_Place.cs:480]: https://github.com/ShokoAnime/ShokoServer/blob/d170da7c7cc219671b10d430da36511aeef0db4d/Shoko.Server/Models/SVR_VideoLocal_Place.cs#L53

I saw this, so I decided to use a "return object" instead of a tuple.

File: Shoko.Server/Renamer/RenameFileHelper.cs (hunk 2 above)
Comment (revam) [reply Shoko.Server/Renamer/RenameFileHelper.cs:29]: Removed.

File: Shoko.Server/Renamer/RenameFileHelper.cs (hunk 1 above)
Comment (revam) [reply Shoko.Server/Renamer/RenameFileHelper.cs:190]: Just tested it and asp.net can handle dynamic routes with strings and static routes just fine.

File: Shoko.Server/Repositories/Direct/RenameScriptRepository.cs (hunk 26 above)
Comment (revam) [reply Shoko.Server/Repositories/Direct/RenameScriptRepository.cs:46]: To avoid checking for and handling null when expecting a list.

File: Shoko.Server/API/v3/Controllers/RenamerController.cs (hunk 16 above)
Comment (Cazzar) [reply Shoko.Server/API/v3/Controllers/RenamerController.cs:187]: unless you're retuning a localisation key, it would be best to make it a human readable error message.

File: Shoko.Server/Models/SVR_VideoLocal_Place.cs (hunk 3 above)
Comment (Cazzar) [reply Shoko.Server/Models/SVR_VideoLocal_Place.cs:401]: But we already have had a working file relocation with feedback in APIv1, what is needing refactor?

Comment (revam) [reply Shoko.Server/Models/SVR_VideoLocal_Place.cs:401]: 1. The method used in the v1 (see below) was also marked as obsolete and returned both the success status and a potential error message, and the error message was not even used in the parts I edited. So I changed it over to the newer implementation (located right below the older implementation).
  https://github.com/ShokoAnime/ShokoServer/blob/1f943396efd4b9695fdf93bf76ee8877097bf840/Shoko.Server/Models/SVR_VideoLocal_Place.cs#L358-L399
2.The newer implementation (this method) did not return an error message but threw instead, and it also did not return success status when it did not throw. So I added the success status to the newer implementation, but looking closer at it now (and comparing the implementations) I can see that was a mistake. It was fine with a void return.

Also, there is feedback in the new v3 relocate endpoint too, it just throws the error (and lets asp.net serialize it) instead of returning a wrapped object with the error message instead.

**Comment by revam** [comment]
API Proposal for the new renamer api.

**Comment by Cazzar** [comment]
I will be working on the renamer changes soon myself, including apiv3

File: Shoko.Server/Models/SVR_VideoLocal_Place.cs (hunk 4 above)
Comment (Cazzar) [reply Shoko.Server/Models/SVR_VideoLocal_Place.cs]: For a simple DTO I would prefer a tuple return as you can very easily destructure them out
